managed/services/backup/removal_service.go

```diff
@@ -27,142 +31,300 @@ import (
-			artifactName+"/"); err != nil {
-			if _, updateErr := models.UpdateArtifact(s.db.Querier, artifactID, models.UpdateArtifactParams{
-				Status: models.BackupStatusPointer(models.FailedToDeleteBackupStatus),
-			}); updateErr != nil {
-				s.l.WithError(updateErr).
-					Errorf("failed to set status %q for artifact %q", models.FailedToDeleteBackupStatus, artifactID)
-			}
+	defer func() {
+		if err != nil {
+			s.setArtifactStatus(artifactID, models.FailedToDeleteBackupStatus)
+		}
+	}()
+
+	restoreItems, err := models.FindRestoreHistoryItems(s.db.Querier, models.RestoreHistoryItemFilters{
+		ArtifactID: artifactID,
+	})
+	if err != nil {
+		return err
+	}
 
+	for _, ri := range restoreItems {
+		if err = models.RemoveRestoreHistoryItem(s.db.Querier, ri.ID); err != nil {
 			return err
 		}
 	}
 
-	return s.db.InTransaction(func(tx *reform.TX) error {
-		restoreItems, err := models.FindRestoreHistoryItems(tx.Querier, models.RestoreHistoryItemFilters{
-			ArtifactID: artifactID,
-		})
+	if removeFiles {
+		go func() {
+			var err error
+			defer func() {
+				if err != nil {
+					s.setArtifactStatus(artifactID, models.FailedToDeleteBackupStatus)
+				}
+			}()
+
+			location, err := models.FindBackupLocationByID(s.db.Querier, artifact.LocationID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't get location")
+				return
+			}
+
+			service, err := models.FindServiceByID(s.db.Querier, artifact.ServiceID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't get service")
+				return
+			}
+
+			if err = s.deleteArtifactFiles(context.Background(), storage, location, artifact, len(artifact.MetadataList)); err != nil {
+				s.l.WithError(err).Error("couldn't delete artifact files")
+				return
+			}
+
+			if service.ServiceType == models.MongoDBServiceType && artifact.Mode == models.PITR {
+				if err = s.deleteArtifactPITRChunks(context.Background(), storage, location, artifact, nil); err != nil {
+					s.l.WithError(err).Error("couldn't delete artifact PITR chunks")
+					return
+				}
+			}
+
+			err = models.DeleteArtifact(s.db.Querier, artifactID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't delete artifact")
+			}
+		}()
+
+		return nil
+	}
+
+	err = models.DeleteArtifact(s.db.Querier, artifactID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TrimPITRArtifact removes first N first records from PITR artifact. Removes snapshots, PITR chunks and corresponding data from database.
+func (s *RemovalService) TrimPITRArtifact(storage Storage, artifactID string, firstN int) error {
+	artifact, oldStatus, err := s.lockArtifact(artifactID, models.RetentionInProgressBackupStatus)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		var err error
+		defer func() {
+			if err != nil {
+				s.l.Error("Couldn't trim artifact files. Restoring is not guaranteed for files outside of retention policy limit.")
+				s.setArtifactStatus(artifactID, models.SuccessBackupStatus)
```

Comment:
This line releases artifact?

Reply:
Right. Logic was changed after one of refactoring cycle. So now changed the func that sets status.